### DISCUSSION TITLE
What is the most lower UEFI network protocol that iPXE provides?

### DISCUSSION THREAD

**Discussion by Kostr** [ipxe/ipxe#1020] [Q&A]
As I understand, vendor OptionROM for network cards usually populate just the `EFI_NETWORK_INTERFACE_IDENTIFIER_PROTOCOL` (i.e pointer to `UNDI iPXE` structure). Based on that edk2-based firmware automatically generates all the other network stack protocols.

```

NII -> SNP -> MNP -> Arp -> IPv4 -> Udp4 -> Dhcp4  --> PxeBc
                                         -> Mtftp4 /
                                         -> Dns
                                 -> Tcp4 -> Http
```

What is the iPXE approach? Which protocols does it generate itself, and which does it expect to be generated from the UEFI firmware (after the iPXE image load)

**Comment by mcb30** [comment]
When using iPXE as a driver, it provides NII and SNP which can be consumed by other UEFI firmware drivers to create MNP, etc.

When using iPXE as an application (which is the usual use case for iPXE), the NII and SNP interfaces will be marked as internally claimed and unavailable for use by other UEFI drivers.  iPXE uses its own (much, much faster) TCP/IP stack, HTTP client, etc.

**Reply by Kostr** [reply]
Thanks!

What is the point to provide both NII and SNP in the first case? Isn't NII is enough, since SNP can be just created in edk2 from NII?

As for the second case. Can we return from the iPXE application to the UEFI shell? Will we still be able to use network stack from shell?

**Reply by mcb30** [reply]
> What is the point to provide both NII and SNP in the first case? Isn't NII is enough, since SNP can be just created in edk2 from NII?

SNP is a relatively sane design, and support for this was implemented first.  The UNDI/NII API design is batshit crazy, and nominal support for providing this API was added only because it would have been even more work to convince the OEM in question that it was a stupid idea to do so.  iPXE does technically provide the UNDI/NII API, but locks it by opening its own protocol instance `BY_DRIVER|EXCLUSIVE` to prevent `SnpDxe` from actually using it.

> As for the second case. Can we return from the iPXE application to the UEFI shell? Will we still be able to use network stack from shell?

If you return from iPXE as an application, then iPXE will tidy up after itself (which includes uninstalling any handles that were created).

You can use iPXE to download and run the UEFI shell, in which case the iPXE device handles will be present from within that shell.

You can also build iPXE as a driver (e.g. `ipxe.efidrv`) rather than as an application (e.g. `ipxe.efi`).  That driver can be loaded from the UEFI shell using the shell's `load` and `connect` commands.
